admin-dev/themes/new-theme/js/pages/product/components/categories/category-tree-selector.ts

```
@@ -0,0 +1,471 @@
+
+      fieldset.classList.remove('d-none');
+      loader.classList.add('d-none');
+    }
+  }
+
+  /**
+   * Used to recursively create items of the category tree
+   */
+  private generateCategoryTree(treeCategory: TreeCategory): HTMLElement {
+    const categoryNode = this.generateTreeElement(treeCategory) as HTMLElement;
+    const childrenList = categoryNode.querySelector(ProductCategoryMap.childrenList) as HTMLElement;
+    childrenList.classList.add('d-none');
+
+    const hasChildren = treeCategory.children && treeCategory.children.length > 0;
+    categoryNode.classList.toggle('more', hasChildren);
+    if (hasChildren) {
+      const inputsContainer = categoryNode.querySelector(ProductCategoryMap.treeElementInputs) as HTMLElement;
+
+      inputsContainer.addEventListener('click', (event) => {
+        // We don't want to mess with the inputs behaviour (no toggle when checkbox or radio is clicked)
+        // So we only toggle when the div itself is clicked.
+        if (event.target !== event.currentTarget) {
+          return;
+        }
+
+        const isExpanded = !childrenList.classList.contains('d-none');
+        categoryNode.classList.toggle('less', !isExpanded);
+        categoryNode.classList.toggle('more', isExpanded);
+        childrenList.classList.toggle('d-none', isExpanded);
+      });
+
+      // Recursively build the children trees
+      treeCategory.children.forEach((childCategory) => {
+        const childTree = this.generateCategoryTree(childCategory);
+
+        childrenList.append(childTree);
+      });
+    }
+
+    return categoryNode;
+  }
+
+  /**
+   * If the category is among the initial ones (inserted by the form on load) the existing element is used,
+   * if not then it is generated based on the prototype template. In both case the element is injected with the
+   * category name.
+   */
+  private generateTreeElement(category: TreeCategory): HTMLElement {
+    if (!this.prototypeTemplate || !this.prototypeName) {
```

You could throw this error in the contructor

---

I moved selection of these templates to generateTreeElement method as it was the only place used. Btw I cannot set them in constructor, because it only appears in DOM on initModal() method. Most of those params are kinda statefull (appears only after initModal()), which I don't like, but idk how to make it better :(